Events view draft

maciaszczykm opened the pull request:
Connected to https://github.com/kubernetes/dashboard/issues/104 issue. Implemented initial version of events view with @floreks. Still a lot of mocks inside, but we will continue to work on that.

Backend loads all event for replica set and pods connected to it using labels. I'm not yet 100% sure it's good solution, but it seems to be working as expected.

@floreks implemented left side sidebar, I was working on events view and partially updated replica set detail view.

Moreover, from important things - updated Angular Material Design to 0.12.0-rc1 version.

maciaszczykm commented:
Regarding the Angular Material Design version update, I see, that there is already pull request https://github.com/kubernetes/dashboard/pull/126, which do the same, but to newer version. If it will be merged before this, we may need to rebase and fix some UI connected problems.

bryk commented on src/app/backend/events.go:
```diff
@@ -72,8 +73,27 @@ func GetEvents(client *client.Client, namespace string) (*Events, error) {
-	for _, element := range list.Items {
-		events.Events = append(events.Events, Event{
+	// Get events for replica set.
+	selector, _ := fields.ParseSelector("involvedObject.name=" + replicaSet.Name)
```
You ignore error here. Just add `if err != nil {` boilerplate everywhere. It is easy. With that we'll not forget about this in the future.

bryk replied:
Try to remove "involvedObject.name" magic string. See how kubectl does this kubernetes/pkg/kubectl/describe.go

```go
func (d *PodDescriber) Describe(namespace, name string) (string, error) {
	rc := d.ReplicationControllers(namespace)
	pc := d.Pods(namespace)

	pod, err := pc.Get(name)
	if err != nil {
		eventsInterface := d.Events(namespace)
		selector := eventsInterface.GetFieldSelector(&name, &namespace, nil, nil)
		options := unversioned.ListOptions{FieldSelector: unversioned.FieldSelector{selector}}
		events, err2 := eventsInterface.List(options)
		if err2 == nil && len(events.Items) > 0 {
			return tabbedString(func(out io.Writer) error {
				fmt.Fprintf(out, "Pod '%v': error '%v', but found events.\n", name, err)
				DescribeEvents(events, out)
				return nil
			})
		}
		return "", err
	}
```

bryk commented on src/app/frontend/index.scss:
```diff
@@ -17,4 +17,4 @@
 
 // Angular Material icon font. It allows to use the icons as fonts.
-@import url(//fonts.googleapis.com/icon?family=Material+Icons);
+@import url(//fonts.googleapis.com/icon?family=Material+Icons);
```
Revert?

bryk commented on src/app/frontend/index_config.js:
```diff
@@ -17,13 +17,23 @@
  * @ngInject
  */
 export default function config($mdThemingProvider) {
+  let kubeBlue = '326de6';
```
I see that the palette configuration object is the same. How about extracting it here to a common variable, instead of just the kubeBlue color?

bryk commented:
Wow, this is a lot of work. You know what? Let's polish the and do not add more features to it until it is merged. Then, when it is merged, we can split the work and pixel-perfect it. 

Always try to send small pull requests and often. With that we can have high quality code reviews and parallel our work. 

<I'm reviewing>

bryk commented on src/app/backend/events.go:
```diff
@@ -72,8 +73,27 @@ func GetEvents(client *client.Client, namespace string) (*Events, error) {
+	labelSelector := labels.SelectorFromSet(replicaSet.Spec.Selector)
+	pods, _ := client.Pods(namespace).List(labelSelector, fields.Everything())
+	for _, pod := range pods.Items {
+		selector, _ := fields.ParseSelector("involvedObject.name=" + pod.Name)
```
Ditto

maciaszczykm replied:
Already done.

bryk commented on src/app/backend/replicasetdetail.go:
```diff
@@ -73,13 +81,33 @@ func GetReplicaSetDetail(client *client.Client, namespace string, name string) (
+		return nil, err
+	}
+
+	matchingServices := getMatchingServices(services.Items, replicaSet)
```
I see similar code block in replica set details. Reuse?

floreks replied:
This part is taken and adjusted from replica set list in order to add details about internal/external endpoints. Should it be changed?

bryk replied:
I mean, you can create a function that returns list of endpoints for service list. And then use this method in two places.

floreks replied:
I did not get that. I will refactor the code.

bryk commented on src/app/frontend/index_config.js:
```diff
@@ -17,13 +17,23 @@
+  $mdThemingProvider.definePalette(kubernetesAccentPaletteName, kubernetesAccentPalette);
+  $mdThemingProvider.theme('default')
+      .primaryPalette(kubernetesColorPaletteName)
+      .accentPalette(kubernetesAccentPaletteName);
```
Nice, thanks!

maciaszczykm commented:
Okay, next time we'll try to split work into smaller pieces. This time all of it was connected so it was harder task, so we went easier path. Moreover, I was working together with @floreks so we were able to do more working at the same time.

I've fixed the issues and rebased. PTAL

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.html:
```diff
@@ -14,27 +14,145 @@
+          <tbody>
+          <tr ng-repeat="event in ctrl.replicaSetEvents.events">
+            <td class="kd-replicasetdetail-table-dark">
+              {{ event.message | limitTo: 50 }}{{event.message.length > 50 ? '...' : ''}}
```
You can do this with pure CSS. See text overload or how replica set list implemented this.

maciaszczykm replied:
Yes, I will refactor it. There is still a lot to do with table styling, most of all with cells size.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.html:
```diff
@@ -14,27 +14,145 @@
+            <td class="kd-replicasetdetail-table-dark">
+              {{ event.message | limitTo: 50 }}{{event.message.length > 50 ? '...' : ''}}
+            </td>
+            <td>{<!>{{event.sourceComponent}} {{event.sourceHost}}<!>}</td>
```
What is this `<!>`

maciaszczykm replied:
I've added it to escape `{` and `}`, which are displayed in view. When `{{{` or `}}}` are used Angular raises some kind of errors, so I have splitted them by `<!>` comments. Is there a better solution for it?

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.html:
```diff
@@ -14,27 +14,145 @@
+          <thead>
+          <tr>
+            <th>Message
+              <i class="material-icons">help</i></th>
```
This needs better formatting.

maciaszczykm replied:
Fixed.

bryk commented on bower.json:
```diff
@@ -8,7 +8,7 @@
     "angular-animate": "~1.4.2",
     "angular-aria": "~1.4.2",
-    "angular-material": "~0.11.4",
+    "angular-material": "~0.12.0-rc1",
```
There are errors on the console on the replica set list page. Remove flex attr from one of the buttons.

bryk replied:
I was wrong. This was caused by my stale bower install. Ignore.

bryk commented on src/app/backend/replicasetdetail.go:
```diff
@@ -46,6 +46,14 @@ type ReplicaSetDetail struct {
+
+	// Internal endpoints of all Kubernetes services have the same label selector as this Replica Set.
+	// Endpoint is DNS name merged with ports.
+	InternalEndpoints []string `json:"internalEndpoints"`
```
I think this model needs a submodel for the service. I.e.., replicaSetDetail contains service information. Replica list contain only endpoints, but that's all they're going to display. Here we have detailed information and label selector for the service. Later we'll add name and possibly other things. I think it is worth to create a nested or named struct here.

What do you think?

floreks replied:
With more information from service it is reasonable to create dedicated structure. We can create ServiceDetail and place all that in there.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.html:
```diff
@@ -14,27 +14,145 @@
+        </h1>
+        <h1 class="md-body-2">Service</h1>
+        <h1 class="md-caption kd-font-1">Label selector</h1>
+        <h1 class="md-caption kd-font-2"><b>{{ctrl.formatLabelString(ctrl.replicaSetDetail.labels)}}</b></h1>
```
This is not true. You should get and display the label selector of the service.

floreks replied:
Changed

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.html:
```diff
@@ -14,27 +14,145 @@
+      <div flex layout="column">
+        <h3 layout="row" class="kd-primary md-body-2">
+          <div flex="55">
+            <i class="kd-align-mid material-icons">refresh</i>
```
I is for italic text, not info icon. Use span for inline elements and div for blocks, if there's nothing better.

floreks replied:
All the icon examples on material icons are within `<i>` tag. Is there a difference if we use span instead of `i` for icons?

bryk replied:
Hmm... If the official spec says so. Let's keep that in mind and add to a
style guide when we have one.


On Fri, Dec 4, 2015 at 2:32 PM, Sebastian Florek <notifications@github.com>
wrote:

> In src/app/frontend/replicasetdetail/replicasetdetail.html
> <https://github.com/kubernetes/dashboard/pull/127#discussion_r46681410>:
>
> > -    </table>
> > +<md-content layout="row" layout-fill flex>
> > +  <div class="kd-replicasetdetail-sidebar" layout="column">
> > +    <div flex layout-align="start center" layout="row" class="kd-header-option-panel">
> > +      <a flex-offset="10" ui-sref="replicasets">
> > +        <i class="material-icons kd-align-mid kd-font-lg">keyboard_backspace</i>
> > +      </a>
> > +      <span flex-offset="5" flex class="kd-align-mid md-title">
> > +        {{ctrl.replicaSetDetail.name}}
> > +      </span>
> > +    </div>
> > +    <div flex-offset="5" flex layout-align="start start" layout-padding layout="row">
> > +      <div flex layout="column">
> > +        <h3 layout="row" class="kd-primary md-body-2">
> > +          <div flex="55">
> > +            <i class="kd-align-mid material-icons">refresh</i>
>
> All the icon examples on material icons are within <i> tag. Is there a
> difference if we use span instead of i for icons?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/dashboard/pull/127/files#r46681410>.
>

bryk replied:
So: keep `<i>`

On Fri, Dec 4, 2015 at 2:32 PM, Sebastian Florek <notifications@github.com>
wrote:

> In src/app/frontend/replicasetdetail/replicasetdetail.html
> <https://github.com/kubernetes/dashboard/pull/127#discussion_r46681410>:
>
> > -    </table>
> > +<md-content layout="row" layout-fill flex>
> > +  <div class="kd-replicasetdetail-sidebar" layout="column">
> > +    <div flex layout-align="start center" layout="row" class="kd-header-option-panel">
> > +      <a flex-offset="10" ui-sref="replicasets">
> > +        <i class="material-icons kd-align-mid kd-font-lg">keyboard_backspace</i>
> > +      </a>
> > +      <span flex-offset="5" flex class="kd-align-mid md-title">
> > +        {{ctrl.replicaSetDetail.name}}
> > +      </span>
> > +    </div>
> > +    <div flex-offset="5" flex layout-align="start start" layout-padding layout="row">
> > +      <div flex layout="column">
> > +        <h3 layout="row" class="kd-primary md-body-2">
> > +          <div flex="55">
> > +            <i class="kd-align-mid material-icons">refresh</i>
>
> All the icon examples on material icons are within <i> tag. Is there a
> difference if we use span instead of i for icons?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/dashboard/pull/127/files#r46681410>.
>

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.scss:
```diff
@@ -0,0 +1,80 @@
+  min-width: 310px;
+}
+
+.kd-replicasetdetail-table td,
```
Do not style th or td. Just add a class to specific elements. With that you'll be able to remove !important from the next rule.

maciaszczykm replied:
Fixed.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.scss:
```diff
@@ -0,0 +1,80 @@
+}
+
+.kd-replicasetdetail-table-dark {
+  color: #000000 !important;
```
See my previous comment.

maciaszczykm replied:
Fixed.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.scss:
```diff
@@ -0,0 +1,80 @@
+  font-size: 1.2em;
+}
+
+md-input-container {
```
Can you use class instead of element name? Input container should have an interesting class

maciaszczykm replied:
Fixed.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.html:
```diff
@@ -14,27 +14,145 @@
+          <b>{{ctrl.formatLabelString(ctrl.replicaSetDetail.labelSelector)}}</b>
+        </h1>
+        <h1 class="md-body-2">Replica Sets</h1>
+        <h1 class="md-caption kd-font-1">Labels</h1>
```
I see that you're using h1-s and apply class md-caption to every row. Instead, it is better to place your visual elements into semantic boxes. For example:
```html
<div class="my-detail">
   <span class="my-detail-title">Label selector</span>
  <span class="my-detail-line">{{ctrl.formatLabelString(ctrl.replicaSetDetail.labelSelector)}}</span>
</div>
```
... and repeat this for all parts. 

With this you dont tie yourself to implementation details ("kd-font-1"). You just say what things are in real life and then style them in css.

floreks replied:
The reason why i'm using headers(h1, h2, etc) is that they have padding(top, bottom) added by default to them and combined with `md-body` `md-caption` i don't have to write css to add vertical spacing myself.

Also I was thinking about font coloring because I've done investigation if this can be tied to mdThemingProvider and it seems that there is no option to do that now. There were few workarounds for that, but they involve some custom services. `kd-font-1` class is in actually in `my-detail` css file but to have consistent font colors across project we could extract that somewhere higher.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.html:
```diff
@@ -14,27 +14,145 @@
+        <h1 class="md-caption kd-font-2"><b>{{ctrl.formatLabelString(ctrl.replicaSetDetail.labels)}}</b></h1>
+        <h1 class="md-caption kd-font-1">Images</h1>
+        <h1 class="md-caption kd-font-2" ng-repeat="image in ctrl.replicaSetDetail.containerImages">
+          <b>{{image}}</b>
```
Dont use `<b>`. Apply this in CSS styles.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.html:
```diff
@@ -14,27 +14,145 @@
+    </div>
+    <div flex-offset="5" flex layout-align="start start" layout-padding layout="row">
+      <div flex layout="column">
+        <h3 layout="row" class="kd-primary md-body-2">
```
Why did you use md-body-2? Can you remove it?

bryk replied:
Hmm... h3 is for content heading. I dont think this is content heading. This is button bar.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.scss:
```diff
@@ -0,0 +1,80 @@
+}
+
+.kd-font-1 {
+  color: #AAAAAA;
```
I see that this CSS contains like 5 colors. Can you reduce their number? If not, extract to constants to give meaning?

floreks replied:
I've proposed solution for that.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail.scss:
```diff
@@ -0,0 +1,80 @@
+  color: #888888;
+}
+
+.kd-editpen-md {
```
`-md` and `-sm` suffixes are used by material to indicate rules that work on specified screen sizes. You should avoid using it unless you do the same.

maciaszczykm replied:
Fixed.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail_controller.js:
```diff
@@ -20,12 +20,38 @@
+   * @return {string}
+   * @export
+   */
+  formatLabelString(object) {
```
Can you reuse this in the details page?

floreks replied:
I can extract this to export function then import in second controller and assign to fn variable. Is it ok or do you have something else in mid?

```javascript
import {formatLabelString} from 'replicasetdetail/replicasetdetail_controller';
.... in controller
this.formatLabelString = formatLabelString;
```

bryk replied:
Hmm... Maybe just add a TODO in both places to keep this PR short.

bryk commented on src/app/frontend/replicasetdetail/replicasetdetail_state.js:
```diff
@@ -67,3 +68,16 @@ export function resolveReplicaSetDetails($stateParams, $resource) {
+ * @ngInject
+ */
+export function resolveReplicaSetEvents($stateParams, $resource) {
+  /** @type {!angular.Resource<!backendApi.ReplicaSetDetail>} */
```
This has wrong type

maciaszczykm replied:
Fixed.

bryk commented:
Nice work, PTAL.

maciaszczykm commented:
PTAL

maciaszczykm commented:
PTAL

bryk commented on src/app/assets/styles/_color_variables.scss:
```diff
@@ -0,0 +1,25 @@
+$subline: #888888;
+
+// Replica Set Details related colors
+$replicasetdetails-sidebar-bg: #FAFAFA;
```
Move all things that are specific only to replica set details page to its css. Follow the rule: "I want to be able to delete replica set detail feature folder to delete all things related to it." 

Also, I'm not sure whether location of this file is correct and its purpose. Please add a todo and create an issue to decide where to put global style variables and how to generally do global css changes.

bryk commented:
Last comment and let's merge this

maciaszczykm commented:
Done. PTAL

floreks commented:
Related theming issue #139
